modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/DefaultSubmosaicProducerFactory.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * GeoTools - The Open Source Java GIS Toolkit
+ * http://geotools.org
+ *
+ * (C) 2016, Open Source Geospatial Foundation (OSGeo)
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation;
+ * version 2.1 of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ */
+
+package org.geotools.gce.imagemosaic.granulecollector;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.geotools.gce.imagemosaic.MergeBehavior;
+import org.geotools.gce.imagemosaic.RasterLayerRequest;
+import org.geotools.gce.imagemosaic.RasterLayerResponse;
+import org.geotools.gce.imagemosaic.RasterManager;
+import org.geotools.resources.coverage.FeatureUtilities;
+import org.opengis.filter.Filter;
+
+/**
+ * Create granule collectors based on whether the request was for a stacked mosaic with additional domains requested
+ */
+public class DefaultSubmosaicProducerFactory implements SubmosaicProducerFactory {
+
+    @Override
+    public List<SubmosaicProducer> createCollectors(RasterLayerRequest request,
+            RasterManager rasterManager, RasterLayerResponse response, boolean dryRun) {
+        // get merge behavior as per request
+        List<SubmosaicProducer> defaultSubmosaicProducers = new ArrayList<>();
+        MergeBehavior mergeBehavior = request.getMergeBehavior();
+
+        // prepare dimensions management if needed, that is in case we use stacking
+        if (mergeBehavior.equals(MergeBehavior.STACK)) {
```

Comment:
Given that we plan to refactor the mosaic in the future to simplify its structure further, I'd say to let this one go for the moment. We are also considering moving the stacking approach up in the food chain and have it implemented purely in GeoServer later (as it's only used by WMS-EO at the moment).